{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUST00000104605.4', 'chromosome:GRCm39:Y:1308273:1308377:-1', 'gene:ENSMUSG00000077793.4', 'gene_biotype:snRNA', 'GGGTTCACTAGGTAGCAAATATACTAAATTTGGAACTATAAAGACAAGATTAGCATGACCCTTGCACAAAATGACATACAAATTAATAAAACATTCCATACTTTT']\n",
      "['ENSMUST00000122627.3', 'chromosome:GRCm39:16:89248383:89248487:-1', 'gene:ENSMUSG00000084576.3', 'gene_biotype:snRNA', 'GTGCTTGCTTAAGCAGCATGCATATTAAAATTGGATCGCCACAGTAGATTAGCATGGCCTCTGAACAAGGCTGACATACAAGTTTTTGAAAAATTTTATTTTTCT']\n",
      "['ENSMUST00000157602.3', 'chromosome:GRCm39:18:55312912:55312991:1', 'gene:ENSMUSG00000088227.3', 'gene_biotype:snRNA', 'GTGCTCGCTTCCGCAGCACATATACTAAAATTGGAACAATACAGAGATTAGCATGGCTCCTGTGCAAAGATGACCCGCAA']\n",
      "['ENSMUST00000083254.3', 'chromosome:GRCm39:16:37800449:37800588:1', 'gene:ENSMUSG00000065188.3', 'gene_biotype:snRNA', 'ATACTTGGCAGGGGAGACACCAATATCAGGAAGAAGGTTTTCCCAGGGTAAGGCTTACCCTGTGCATTTGGGGTGTGCTGACCCCTGAGTAAATTTGACTTCATAATTTGTGGTAGTGAGGAGTACATTTGAGCTCTTTC']\n",
      "['ENSMUST00020182548.1', 'chromosome:GRCm39:18:20977504:20977649:-1', 'gene:ENSMUSG00002076630.1', 'gene_biotype:snRNA', 'ATACTTAACCTGACAGGGATACACCATGATTAAGAAAGTGATCTTTCCAGGGCAAGACTCTTCCATTGCACTTGGGATTTGCTGACCTAAAATATACTTTCACTAAATGGGAGAAATGCCACTGAAGATTCTGTGATAAAATTTGA']\n"
     ]
    }
   ],
   "source": [
    "from Bio import TogoWS, SeqIO\n",
    "# FASTAファイルを読み込む\n",
    "file_path = \"/Users/takumu/VScode/Mus_musculus_db/Mus_musculus.GRCm39.ncrna.fa\"\n",
    "records = list(SeqIO.parse(file_path, \"fasta\"))\n",
    "\n",
    "# 各エントリに関する情報を取得する\n",
    "rna_info = []\n",
    "for record in records:\n",
    "    y = []\n",
    "    x= list(record.description.split(\" \")) # エントリの情報をスペースで区切ってリストに格納\n",
    "    y.append(x[0]) # エントリIDをリストに格納\n",
    "    y.append(x[2]) \n",
    "    y.append(x[3])\n",
    "    y.append(x[4])\n",
    "    y.append(str(record.seq))\n",
    "    rna_info.append(y)\n",
    "\n",
    "# 先頭5つのエントリに関する情報を表示\n",
    "for info in rna_info[:5]:\n",
    "    print(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00020182969.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>20134948</td>\n",
       "      <td>20135057</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002074919.1</td>\n",
       "      <td>GGGCATCTGCTGACATGGGGGCCGAAGTCAGTCGCCAGGCAGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00020183589.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>991657</td>\n",
       "      <td>991716</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSMUSG00002076920.1</td>\n",
       "      <td>TCTCATCTGATCTCGGAAGCTAAGCAGGGTCGGGCCTGGTTAGTAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00020183415.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>2446010</td>\n",
       "      <td>2446108</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSMUSG00002075837.1</td>\n",
       "      <td>TTAACTTTTACTTTAATAACATATATATGTATGTGTGTGTGTGTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00020183034.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>5542752</td>\n",
       "      <td>5542861</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002075850.1</td>\n",
       "      <td>GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00020181908.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>40639526</td>\n",
       "      <td>40639635</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002075449.1</td>\n",
       "      <td>GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUST00020182429.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>40644083</td>\n",
       "      <td>40644192</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002074974.1</td>\n",
       "      <td>GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGCGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUST00020181878.1</td>\n",
       "      <td>18</td>\n",
       "      <td>18232617</td>\n",
       "      <td>18232717</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00002076588.1</td>\n",
       "      <td>GAAATGCCTCCATGAGATCCAACTGTAAGGCATTTTCTCAATTAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUST00020182316.1</td>\n",
       "      <td>16</td>\n",
       "      <td>13946430</td>\n",
       "      <td>13946495</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSMUSG00002076089.1</td>\n",
       "      <td>GACTGGGATGTCACGGAATATGGTGTTTGTGACTTGATTGATATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUST00000093566.3</td>\n",
       "      <td>16</td>\n",
       "      <td>13977490</td>\n",
       "      <td>13977556</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000070074.3</td>\n",
       "      <td>CTCGTCAGGCTCAGTCCCCTCCCGATAAACCTCAAAATAGGGTCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUST00020182722.1</td>\n",
       "      <td>18</td>\n",
       "      <td>56099924</td>\n",
       "      <td>56100013</td>\n",
       "      <td>-1</td>\n",
       "      <td>ENSMUSG00002076978.1</td>\n",
       "      <td>TGTGTGTGTGTGTGTGTGTGTGTGTGTAATAGCTATATATGTACAC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID chromosome     start       end strand  \\\n",
       "0  ENSMUST00020182969.1          Y  20134948  20135057      1   \n",
       "1  ENSMUST00020183589.1          Y    991657    991716     -1   \n",
       "2  ENSMUST00020183415.1          Y   2446010   2446108     -1   \n",
       "3  ENSMUST00020183034.1          Y   5542752   5542861      1   \n",
       "4  ENSMUST00020181908.1          Y  40639526  40639635      1   \n",
       "5  ENSMUST00020182429.1          Y  40644083  40644192      1   \n",
       "6  ENSMUST00020181878.1         18  18232617  18232717      1   \n",
       "7  ENSMUST00020182316.1         16  13946430  13946495     -1   \n",
       "8  ENSMUST00000093566.3         16  13977490  13977556      1   \n",
       "9  ENSMUST00020182722.1         18  56099924  56100013     -1   \n",
       "\n",
       "                   gene                                                seq  \n",
       "0  ENSMUSG00002074919.1  GGGCATCTGCTGACATGGGGGCCGAAGTCAGTCGCCAGGCAGCGGG...  \n",
       "1  ENSMUSG00002076920.1  TCTCATCTGATCTCGGAAGCTAAGCAGGGTCGGGCCTGGTTAGTAC...  \n",
       "2  ENSMUSG00002075837.1  TTAACTTTTACTTTAATAACATATATATGTATGTGTGTGTGTGTGT...  \n",
       "3  ENSMUSG00002075850.1  GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGTGGG...  \n",
       "4  ENSMUSG00002075449.1  GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGCGGG...  \n",
       "5  ENSMUSG00002074974.1  GGGCATCTGCTGACATGGGGGCCGAAGTCAGGCGCCAGGCAGCGGG...  \n",
       "6  ENSMUSG00002076588.1  GAAATGCCTCCATGAGATCCAACTGTAAGGCATTTTCTCAATTAGT...  \n",
       "7  ENSMUSG00002076089.1  GACTGGGATGTCACGGAATATGGTGTTTGTGACTTGATTGATATAA...  \n",
       "8  ENSMUSG00000070074.3  CTCGTCAGGCTCAGTCCCCTCCCGATAAACCTCAAAATAGGGTCTT...  \n",
       "9  ENSMUSG00002076978.1  TGTGTGTGTGTGTGTGTGTGTGTGTGTAATAGCTATATATGTACAC...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "rna_df = pd.DataFrame(rna_info,columns=[\"ID\",\"describe\",\"gene\",\"biotype\",\"seq\"])\n",
    "miRNA_1 = rna_df[rna_df.columns.values][rna_df[\"biotype\"] == \"gene_biotype:miRNA\"]\n",
    "describe = miRNA_1[\"describe\"].values\n",
    "describe = [x.split(\":\") for x in describe]\n",
    "describe = describe[5:]\n",
    "chromosome = [x[-4] for x in describe]\n",
    "start = [x[-3] for x in describe]\n",
    "end = [x[-2] for x in describe]\n",
    "strand = [x[-1] for x in describe]\n",
    "id = miRNA_1[\"ID\"].values\n",
    "id = id[5:]\n",
    "seq = miRNA_1[\"seq\"].values\n",
    "seq = seq[5:]\n",
    "gene = miRNA_1[\"gene\"].values\n",
    "gene = [x.split(\":\") for x in gene]\n",
    "gene= [x[1] for x in gene[5:]]\n",
    "miRNA = pd.DataFrame(id,columns=[\"ID\"])\n",
    "miRNA[\"chromosome\"] = chromosome\n",
    "miRNA[\"start\"] = start\n",
    "miRNA[\"end\"] = end\n",
    "miRNA[\"strand\"] = strand\n",
    "miRNA[\"gene\"] = gene\n",
    "miRNA[\"seq\"] = seq \n",
    "miRNA.to_csv(\"miRNA.csv\",index=False)\n",
    "miRNA.head(10)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
